scrapy/core/engine.py

```diff
@@ -349,12 +366,26 @@ def errback(failure):
 
         return dfd
 
-    def _close_all_spiders(self):
-        dfds = [self.close_spider(s, reason='shutdown') for s in self.open_spiders]
-        dlist = defer.DeferredList(dfds)
-        return dlist
+    @property
+    def open_spiders(self) -> list:
+        warnings.warn(
+            "ExecutionEngine.open_spiders is deprecated, please use ExecutionEngine.spider instead",
+            category=ScrapyDeprecationWarning,
+            stacklevel=2,
+        )
+        return [self.spider] if self.spider is not None else []
 
-    @defer.inlineCallbacks
-    def _finish_stopping_engine(self):
-        yield self.signals.send_catch_log_deferred(signal=signals.engine_stopped)
-        self._closewait.callback(None)
+    def has_capacity(self) -> bool:
+        warnings.warn("ExecutionEngine.has_capacity is deprecated", ScrapyDeprecationWarning, stacklevel=2)
+        return not bool(self.slot)
+
+    def schedule(self, request: Request, spider: Spider) -> None:
+        warnings.warn(
+            "ExecutionEngine.schedule is deprecated, please use "
+            "ExecutionEngine.crawl or ExecutionEngine.download instead",
+            category=ScrapyDeprecationWarning,
+            stacklevel=2,
+        )
+        if self.slot is None:
+            raise RuntimeError("Engine slot not assigned")
+        self._schedule_request(request)
```

Comment:
A potential issue I see here and in public methods where the `spider` argument is deprecated is that there is a slight behavior change that, while unlikely to affect many, if any, still happens without a deprecation period: we are just assuming that the `spider` argument, if passed, is the same as `self.spider`. And while that is what _should_ be happening, code-wise we cannot be sure of that.

So I’m thinking that, to deprecate things properly, we may need to add a check specifically for that, and have things work as before in those cases (i.e. pass the `spider` argument forward to subsequent calls, instead of letting `self.spider` be used) but log a deprecation warning about it.

Reply:
Makes sense, I added the arguments back to the underscored methods that are called from public ones.
Note that this doesn't include `_spider_idle`, `_next_request`, `_next_request_from_scheduler`. `_spider_idle` and `_next_request_from_scheduler` are only called from `_next_request`, which is in turn only "called" (set to be called later, that is) from `open_spider`, the method which sets `self.spider`.
This should make things safer. Nobody should be calling `crawl/download` with other spiders though, I'm not even sure what would happen but it's probably not pretty.

Reply:
Actually, the only risk is with `download`: `spider_is_idle` was not using the spider to begin with and `crawl` was checking if the spider was in `self.open_spiders` (now it checks if it matches `self.spider`). As I mentioned in https://github.com/scrapy/scrapy/pull/5090#discussion_r614277741, I added the same check to `download`.